File: class1_ROS_Basic/act_sercli_example_pkg/src/simple_action_client_no_wait.py
#! /usr/bin/env python

import rospy
import time
import actionlib
from custom_messages.msg import CounterAction, CounterGoal, CounterResult, CounterFeedback

"""
class SimpleGoalState:
    PENDING = 0
    ACTIVE = 1
    DONE = 2
    WARN = 3
    ERROR = 4

"""
# We create some constants with the corresponing vaules from the SimpleGoalState class
PENDING = 0
ACTIVE = 1
DONE = 2
WARN = 3
ERROR = 4

nImage = 1

# definition of the feedback callback. This will be called when feedback
# is received from the action server
# it just prints a message indicating a new message has been received
def feedback_callback(feedback):
    print("feedback : ",feedback)

def result_callback(status,result):
    print('status : ',status)
    print('result :',result.count)


# initializes the action client node
rospy.init_node('counter_client')

action_server_name = '/counter_server'
client = actionlib.SimpleActionClient(action_server_name, CounterAction)

# waits until the action server is up and running
rospy.loginfo('Waiting for action Server '+action_server_name)
client.wait_for_server()
rospy.loginfo('Action Server Found...'+action_server_name)

# creates a goal to send to the action server
goal = CounterGoal()
goal.max_number = 10
goal.duration = 3
print(goal)

client.send_goal(goal, feedback_cb=feedback_callback, done_cb=result_callback)


# You can access the SimpleAction Variable "simple_state", that will be 1 if active, and 2 when finished.
#Its a variable, better use a function like get_state.
#state = client.simple_state
# state_result will give the FINAL STATE. Will be 1 when Active, and 2 if NO ERROR, 3 If Any Warning, and 3 if ERROR
state_result = client.get_state()

rate = rospy.Rate(1)

rospy.loginfo("state_result: "+str(state_result))

while state_result < DONE:
    rospy.loginfo("Doing Stuff while waiting for the Server to give a result....")
    rate.sleep()
    state_result = client.get_state()
    rospy.loginfo("state_result: "+str(state_result))
    
rospy.loginfo("[Result] State: "+str(state_result))
if state_result == ERROR:
    rospy.logerr("Something went wrong in the Server Side")
if state_result == WARN:
    rospy.logwarn("There is a warning in the Server Side")

File: class1_ROS_Basic/act_sercli_example_pkg_c/src/simple_action_client_no_wait.cpp
#include <ros/ros.h>
#include <custom_messages/CounterAction.h> // Note: "Action" is appended
#include <actionlib/client/simple_action_client.h>

int nImage = 0;

void doneCb(const actionlib::SimpleClientGoalState& state,
            const custom_messages::CounterResultConstPtr& result)
{
  ROS_INFO("result : %d",result->count);
  ros::shutdown();
}

// Called once when the goal becomes active
void activeCb()
{
  ROS_INFO("Goal just went active");
}

void feedbackCb(const custom_messages::CounterFeedbackConstPtr& feedback)
{
  ROS_INFO("feedback : %f", feedback->percentage);
}

int main(int argc, char** argv)
{
  ros::init(argc, argv, "drone_action_client");
  actionlib::SimpleActionClient<custom_messages::CounterAction> client("counter_server", true);
  client.waitForServer();

  custom_messages::CounterGoal goal;
  goal.nseconds = 10;
  
  client.sendGoal(goal, &doneCb, &activeCb, &feedbackCb);
  //client.waitForResult();
  
  ros::Rate loop_rate(2);
  actionlib::SimpleClientGoalState state_result = client.getState();
  ROS_INFO("[State Result]: %s", state_result.toString().c_str());
    
  while ( state_result == actionlib::SimpleClientGoalState::ACTIVE || state_result == actionlib::SimpleClientGoalState::PENDING )
  {
    ROS_INFO("Doing Stuff while waiting for the Server to give a result...");
    loop_rate.sleep();
    state_result = client.getState();
    ROS_INFO("[State Result]: %s", state_result.toString().c_str());
  }
    
  return 0;
}